lib/teleterm/services/desktop/directorysharing.go

```diff
@@ -0,0 +1,296 @@
+				parent := filepath.Dir(full)
+				resolvedParent, perr := filepath.EvalSymlinks(parent)
+				if perr != nil {
+					return "", trace.Wrap(perr)
+				}
+
+				// Reconstruct the full path by joining the resolved parent with the original file name.
+				resolved = filepath.Join(resolvedParent, filepath.Base(full))
+			} else {
+				return "", trace.Wrap(err)
+			}
+		}
+		if !isSubPath(base, resolved) {
+			return "", trace.BadParameter("path escapes from parent")
+		}
+		return resolved, nil
+	}
+	return nil
+}
+
+func isSubPath(parent, child string) bool {
+	return child == parent || strings.HasPrefix(child, parent+string(filepath.Separator))
+}
+
+// Stat retrieves metadata about a file or directory at the given path.
+func (s *DirectoryAccess) Stat(relativePath string) (*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := s.readFileOrDirInfo(relativePath, stat)
+	return info, trace.Wrap(err)
+}
+
+// ReadDir lists files and directories within the given directory path, skips symlinks.
+func (s *DirectoryAccess) ReadDir(relativePath string) ([]*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var results []*FileOrDirInfo
+	for _, entry := range entries {
+		fileInfo, err := entry.Info()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Skip symlinks, we can't present them properly in the remote machine.
+		if fileInfo.Mode().Type()&os.ModeSymlink != 0 {
+			continue
+		}
+
+		entryRelativePath := filepath.Join(relativePath, fileInfo.Name())
+		fileOrDir, err := s.readFileOrDirInfo(entryRelativePath, fileInfo)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		results = append(results, fileOrDir)
+	}
+
+	return results, nil
+}
+
+// Read reads a slice of a file.
+func (s *DirectoryAccess) Read(relativePath string, offset int64, length uint32) ([]byte, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	opened, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer opened.Close()
+
+	buf := make([]byte, length)
+	_, err = opened.Seek(offset, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+
+	n, err := opened.Read(buf)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return nil, trace.Wrap(err)
+	}
+	return buf[:n], nil
+}
+
+// Write writes data to a file at a given offset.
+func (s *DirectoryAccess) Write(relativePath string, offset int64, data []byte) (int, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
```

Comment:
Errors from `Close` shouldn't be ignored. https://www.joeshaw.org/dont-defer-close-on-writable-files/

idk how much truth is in that, but I don't see why it'd be wrong. [I asked on Slack about it](https://gravitational.slack.com/archives/C08HF8E9F/p1747045711129909).

> The simplest way to solve this is simply not to use `defer` when writing files:
> 
> ```go
> func helloNotes() error {
>     f, err := os.Create("/home/joeshaw/notes.txt")
>     if err != nil {
>         return err
>     }
> 
>     if err = io.WriteString(f, "hello world"); err != nil {
>         f.Close()
>         return err
>     }
> 
>     return f.Close()
> }
> ```

In your case it'd mean manually calling `file.Close` when handling errors from `Seek` and `Write`.

Reply:
I saw the Slack discussion and it seems to me that the most ergonomic form is the one with capturing the close error. I applied it everywhere.